modules/fedimint-ecash-migration-client/src/lib.rs

```diff
@@ -57,28 +104,969 @@ impl ClientModule for EcashMigrationClientModule {
+                                    Err(e) => {
+                                        warn!(?operation_id, "Error occurred while waiting for ecash change: {e:?}");
+                                    }
+                                }
+                                yield FundTransferState::Success {
+                                    transfer_id: fund_transfer_success.transfer_id,
+                                    amount: fund_transfer_success.amount,
+                                };
+                                return;
+                            }
+                        }
+                    }
+
+                    unreachable!("State machine update stream ended unexpectedly");
+                }
+            }))
+    }
+
+    /// Upload the origin federation's keyset to the destination federation.
+    ///
+    /// This reads the keyset from the origin config file and uploads it
+    /// to associate with the given liability transfer. The keyset is uploaded
+    /// to all federation peers.
+    ///
+    /// # Errors
+    /// - If the origin config file cannot be read or parsed.
+    /// - If uploading to any peer fails.
+    pub async fn upload_keyset(
+        &self,
+        transfer_id: TransferId,
+        origin_config_path: impl AsRef<Path>,
+    ) -> anyhow::Result<()> {
+        use fedimint_api_client::api::FederationApiExt as _;
+        use fedimint_core::module::ApiRequestErased;
+
+        let tier_keys = read_origin_keyset(origin_config_path).await?;
+
+        let request = UploadKeySetRequest {
+            transfer_id,
+            tier_keys,
+        };
+
+        // Upload to all peers
+        let params = ApiRequestErased::new(request);
+        let results = futures::future::join_all(self.module_api.all_peers().iter().map(|&peer| {
+            let params = params.clone();
+            async move {
+                (
+                    peer,
+                    self.module_api
+                        .request_single_peer::<()>(
+                            UPLOAD_KEY_SET_ENDPOINT.to_string(),
+                            params,
+                            peer,
+                        )
+                        .await,
+                )
+            }
+        }))
+        .await;
+
+        // Check for errors
+        let errors: Vec<_> = results
+            .into_iter()
+            .filter_map(|(peer, result)| result.err().map(|e| (peer, e)))
+            .collect();
+
+        if !errors.is_empty() {
+            anyhow::bail!(
+                "Failed to upload keyset to peers: {:?}",
+                errors
+                    .into_iter()
+                    .map(|(p, e)| format!("{p}: {e}"))
+                    .collect::<Vec<_>>()
+            );
+        }
+
+        Ok(())
+    }
+
+    /// Upload the spend book entries in chunks to the destination federation.
+    ///
+    /// Each chunk is accompanied by a Merkle proof that allows the server to
```

Comment:
Do we really need a merkle tree? Our case seems much simpler. Controllers could just prepare a list of threshold sigs for each batch of nonces, and maybe even we could just hardcode a batch size (to aim at like 1MB chunks), so nonces are uplaoded in signed batches. It's not really server problems to verify a signed batch belongs to the root. If it doesn't, it is sender's problem that they wasted the fee for uploading corrupted data.